{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4f3fad0-5285-48b3-a7f3-225f42593328",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "from pybedtools import BedTool\n",
    "from collections import defaultdict\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e18482ad-e4b9-4ba4-a576-62009ed69106",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = \"test_data_m6anet_mod_sites.csv\"\n",
    "gtf_file = \"gencode.v32.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "83fc2b35-cc39-42fa-ad1f-07b44f588321",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  \n",
       "0               0.980850  GAACT   0.655172        s1     caud  \n",
       "1               0.965541  AGACT   0.913043        s1     caud  \n",
       "2               0.927230  GGACA   0.666667        s1     caud  \n",
       "3               0.936434  GGACT   0.687898        s1     caud  \n",
       "4               0.938970  AAACT   0.567010        s1     caud  \n",
       "5               0.969710  GGACT   0.777778        s2     caud  \n",
       "6               0.949604  GGACT   0.761905        s2     caud  \n",
       "7               0.928761  GGACT   0.761905        s2     caud  \n",
       "8               0.941420  GGACT   0.765432        s2     caud  \n",
       "9               0.903208  AAACT   0.558559        s2     caud  \n",
       "10              0.936208  GGACT   0.779412        s2     caud  \n",
       "11              0.911677  AAACT   0.556701        s2     caud  \n",
       "12              0.987061  GGACT   0.909091        s2     caud  \n",
       "13              0.995830  GGACA   0.962963        s2     caud  \n",
       "14              0.952746  GGACT   0.677419        s2     caud  \n",
       "15              0.956273  AGACA   0.770492        s2     caud  \n",
       "16              0.955948  AGACA   0.794872        s2     caud  \n",
       "17              0.966780  AGACA   0.868421        s2     caud  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Loading data...\")\n",
    "\n",
    "# Load modification sites\n",
    "modification_sites = pd.read_csv(f)\n",
    "modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f76a5429-b8a1-4bd0-bb62-4e11b36e195b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gffutils.interface.FeatureDB at 0x7f0f441a3a40>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_file = gtf_file + \".db\"\n",
    "try:\n",
    "    db = gffutils.FeatureDB(db_file, keep_order=True)\n",
    "except:\n",
    "    print(\"Creating GTF database...\")\n",
    "    db = gffutils.create_db(gtf_file, dbfn=db_file, force=True, keep_order=True, disable_infer_genes=True, disable_infer_transcripts=True)\n",
    "    db = gffutils.FeatureDB(db_file, keep_order=True)\n",
    "db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "da45c5e7-8e85-40ff-b1eb-d50d886b2c82",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([<Feature transcript (chr1:634376-634922[+]) at 0x7f0f43327950>,\n",
       "  <Feature exon (chr1:634376-634922[+]) at 0x7f0f43327e90>,\n",
       "  <Feature transcript (chr1:725885-778626[-]) at 0x7f0f4334cb90>,\n",
       "  <Feature exon (chr1:778284-778626[-]) at 0x7f0f4334ce30>,\n",
       "  <Feature exon (chr1:774171-774280[-]) at 0x7f0f4334c7d0>,\n",
       "  <Feature exon (chr1:772976-773107[-]) at 0x7f0f4334cd40>,\n",
       "  <Feature exon (chr1:769497-769712[-]) at 0x7f0f4334d0d0>,\n",
       "  <Feature exon (chr1:768548-768613[-]) at 0x7f0f4334d310>,\n",
       "  <Feature exon (chr1:766329-766387[-]) at 0x7f0f4334d520>,\n",
       "  <Feature exon (chr1:764723-765247[-]) at 0x7f0f4334d730>,\n",
       "  <Feature exon (chr1:758967-759123[-]) at 0x7f0f4334d940>,\n",
       "  <Feature exon (chr1:756077-756141[-]) at 0x7f0f4334db50>,\n",
       "  <Feature exon (chr1:746695-746818[-]) at 0x7f0f4334dd60>,\n",
       "  <Feature exon (chr1:744728-744825[-]) at 0x7f0f4334df70>,\n",
       "  <Feature exon (chr1:744195-744391[-]) at 0x7f0f4334e1b0>,\n",
       "  <Feature exon (chr1:743180-743350[-]) at 0x7f0f4334e3f0>,\n",
       "  <Feature exon (chr1:735423-735613[-]) at 0x7f0f4334e630>,\n",
       "  <Feature exon (chr1:733307-733364[-]) at 0x7f0f4334e870>,\n",
       "  <Feature exon (chr1:725885-729804[-]) at 0x7f0f4334eab0>,\n",
       "  <Feature transcript (chr1:1216908-1232001[-]) at 0x7f0f4334f1d0>,\n",
       "  <Feature exon (chr1:1231892-1232001[-]) at 0x7f0f4334f350>,\n",
       "  <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f4334f1a0>,\n",
       "  <Feature CDS (chr1:1228468-1228772[-]) at 0x7f0f4334f0e0>,\n",
       "  <Feature start_codon (chr1:1228770-1228772[-]) at 0x7f0f4334f5c0>,\n",
       "  <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4334f7d0>,\n",
       "  <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f4334fa10>,\n",
       "  <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4334fc80>,\n",
       "  <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f4334fec0>,\n",
       "  <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43358170>,\n",
       "  <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f433583b0>,\n",
       "  <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43358620>,\n",
       "  <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f43358860>,\n",
       "  <Feature exon (chr1:1216908-1217688[-]) at 0x7f0f43358ad0>,\n",
       "  <Feature CDS (chr1:1217515-1217688[-]) at 0x7f0f43358d10>,\n",
       "  <Feature stop_codon (chr1:1217512-1217514[-]) at 0x7f0f43358f80>,\n",
       "  <Feature UTR (chr1:1231892-1232001[-]) at 0x7f0f43359190>,\n",
       "  <Feature UTR (chr1:1228773-1228946[-]) at 0x7f0f43359400>,\n",
       "  <Feature UTR (chr1:1216908-1217514[-]) at 0x7f0f43359670>,\n",
       "  <Feature transcript (chr1:1216909-1232067[-]) at 0x7f0f433598e0>,\n",
       "  <Feature exon (chr1:1231892-1232067[-]) at 0x7f0f43359af0>,\n",
       "  <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43359d60>,\n",
       "  <Feature CDS (chr1:1228468-1228793[-]) at 0x7f0f4335a000>,\n",
       "  <Feature start_codon (chr1:1228791-1228793[-]) at 0x7f0f4335a2d0>,\n",
       "  <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4335a540>,\n",
       "  <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f4335a7e0>,\n",
       "  <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4335aab0>,\n",
       "  <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f4335ad50>,\n",
       "  <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f4335b020>,\n",
       "  <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f4335b2c0>,\n",
       "  <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f4335b590>,\n",
       "  <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f4335b830>,\n",
       "  <Feature exon (chr1:1216909-1217804[-]) at 0x7f0f4335bb00>,\n",
       "  <Feature CDS (chr1:1217673-1217804[-]) at 0x7f0f4335bda0>,\n",
       "  <Feature stop_codon (chr1:1217670-1217672[-]) at 0x7f0f433740b0>,\n",
       "  <Feature UTR (chr1:1231892-1232067[-]) at 0x7f0f43374320>,\n",
       "  <Feature UTR (chr1:1228794-1228946[-]) at 0x7f0f433745f0>,\n",
       "  <Feature UTR (chr1:1216909-1217672[-]) at 0x7f0f433748c0>,\n",
       "  <Feature transcript (chr1:1216931-1232031[-]) at 0x7f0f43374b90>,\n",
       "  <Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43374e00>,\n",
       "  <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43374fb0>,\n",
       "  <Feature CDS (chr1:1228468-1228793[-]) at 0x7f0f43375190>,\n",
       "  <Feature start_codon (chr1:1228791-1228793[-]) at 0x7f0f433753a0>,\n",
       "  <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43375550>,\n",
       "  <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f43375730>,\n",
       "  <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43375940>,\n",
       "  <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f43375b20>,\n",
       "  <Feature exon (chr1:1218769-1218943[-]) at 0x7f0f43375d30>,\n",
       "  <Feature CDS (chr1:1218769-1218943[-]) at 0x7f0f43375f10>,\n",
       "  <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43376120>,\n",
       "  <Feature CDS (chr1:1218624-1218633[-]) at 0x7f0f43376300>,\n",
       "  <Feature stop_codon (chr1:1218621-1218623[-]) at 0x7f0f43376510>,\n",
       "  <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f433766c0>,\n",
       "  <Feature UTR (chr1:1231892-1232031[-]) at 0x7f0f433768a0>,\n",
       "  <Feature UTR (chr1:1228794-1228946[-]) at 0x7f0f43376ab0>,\n",
       "  <Feature UTR (chr1:1218458-1218623[-]) at 0x7f0f43376cc0>,\n",
       "  <Feature UTR (chr1:1216931-1217804[-]) at 0x7f0f43376ed0>,\n",
       "  <Feature transcript (chr1:1216931-1232031[-]) at 0x7f0f433770e0>,\n",
       "  <Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43377290>,\n",
       "  <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f433774a0>,\n",
       "  <Feature CDS (chr1:1228468-1228772[-]) at 0x7f0f433776e0>,\n",
       "  <Feature start_codon (chr1:1228770-1228772[-]) at 0x7f0f43377950>,\n",
       "  <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43377b60>,\n",
       "  <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f43377da0>,\n",
       "  <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43377fe0>,\n",
       "  <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f43194290>,\n",
       "  <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43194500>,\n",
       "  <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f43194740>,\n",
       "  <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f431949b0>,\n",
       "  <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f43194bf0>,\n",
       "  <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f43194e60>,\n",
       "  <Feature CDS (chr1:1217673-1217804[-]) at 0x7f0f431950a0>,\n",
       "  <Feature stop_codon (chr1:1217670-1217672[-]) at 0x7f0f43195310>,\n",
       "  <Feature UTR (chr1:1231892-1232031[-]) at 0x7f0f43195520>,\n",
       "  <Feature UTR (chr1:1228773-1228946[-]) at 0x7f0f43195790>,\n",
       "  <Feature UTR (chr1:1216931-1217672[-]) at 0x7f0f43195a00>,\n",
       "  <Feature transcript (chr1:1216932-1221351[-]) at 0x7f0f43195c70>,\n",
       "  <Feature exon (chr1:1218769-1221351[-]) at 0x7f0f43195e80>,\n",
       "  <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43196030>,\n",
       "  <Feature exon (chr1:1216932-1217688[-]) at 0x7f0f43196210>,\n",
       "  <Feature transcript (chr1:1216934-1220293[-]) at 0x7f0f431963f0>,\n",
       "  <Feature exon (chr1:1218458-1220293[-]) at 0x7f0f431965a0>,\n",
       "  <Feature exon (chr1:1216934-1217688[-]) at 0x7f0f43196750>,\n",
       "  <Feature transcript (chr1:1232237-1235041[+]) at 0x7f0f43196cf0>,\n",
       "  <Feature exon (chr1:1232237-1235041[+]) at 0x7f0f43196e10>,\n",
       "  <Feature CDS (chr1:1232279-1233265[+]) at 0x7f0f43196900>,\n",
       "  <Feature start_codon (chr1:1232279-1232281[+]) at 0x7f0f43197080>,\n",
       "  <Feature stop_codon (chr1:1233266-1233268[+]) at 0x7f0f43197320>,\n",
       "  <Feature UTR (chr1:1232237-1232278[+]) at 0x7f0f431975c0>,\n",
       "  <Feature UTR (chr1:1233266-1235041[+]) at 0x7f0f431978c0>,\n",
       "  <Feature transcript (chr1:1292390-1305929[-]) at 0x7f0f43197f80>,\n",
       "  <Feature exon (chr1:1304086-1305929[-]) at 0x7f0f43197ef0>,\n",
       "  <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f43197b90>,\n",
       "  <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b4290>,\n",
       "  <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431b4410>,\n",
       "  <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431b45f0>,\n",
       "  <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431b47d0>,\n",
       "  <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431b49b0>,\n",
       "  <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431b4b90>,\n",
       "  <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431b4d70>,\n",
       "  <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431b4f50>,\n",
       "  <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431b5130>,\n",
       "  <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431b5340>,\n",
       "  <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431b5550>,\n",
       "  <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431b5760>,\n",
       "  <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431b5970>,\n",
       "  <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431b5b80>,\n",
       "  <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431b5d90>,\n",
       "  <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431b5fa0>,\n",
       "  <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431b61b0>,\n",
       "  <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431b63c0>,\n",
       "  <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431b65d0>,\n",
       "  <Feature exon (chr1:1292390-1293708[-]) at 0x7f0f431b67e0>,\n",
       "  <Feature transcript (chr1:1292391-1307930[-]) at 0x7f0f431b69f0>,\n",
       "  <Feature exon (chr1:1307769-1307930[-]) at 0x7f0f431b6bd0>,\n",
       "  <Feature CDS (chr1:1307769-1307815[-]) at 0x7f0f431b6e40>,\n",
       "  <Feature start_codon (chr1:1307813-1307815[-]) at 0x7f0f431b7110>,\n",
       "  <Feature exon (chr1:1304086-1304143[-]) at 0x7f0f431b7380>,\n",
       "  <Feature CDS (chr1:1304086-1304143[-]) at 0x7f0f431b7620>,\n",
       "  <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f431b78f0>,\n",
       "  <Feature CDS (chr1:1303162-1303281[-]) at 0x7f0f431b7b90>,\n",
       "  <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b7e60>,\n",
       "  <Feature CDS (chr1:1302922-1302975[-]) at 0x7f0f431cc140>,\n",
       "  <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431cc410>,\n",
       "  <Feature CDS (chr1:1301988-1302046[-]) at 0x7f0f431cc6b0>,\n",
       "  <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431cc980>,\n",
       "  <Feature CDS (chr1:1300509-1300692[-]) at 0x7f0f431ccc20>,\n",
       "  <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431ccef0>,\n",
       "  <Feature CDS (chr1:1300158-1300202[-]) at 0x7f0f431cd190>,\n",
       "  <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431cd460>,\n",
       "  <Feature CDS (chr1:1299973-1300068[-]) at 0x7f0f431cd700>,\n",
       "  <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431cd9d0>,\n",
       "  <Feature CDS (chr1:1299831-1299905[-]) at 0x7f0f431cdc70>,\n",
       "  <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431cdf40>,\n",
       "  <Feature CDS (chr1:1299345-1299356[-]) at 0x7f0f431ce1e0>,\n",
       "  <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431ce4e0>,\n",
       "  <Feature CDS (chr1:1298567-1298679[-]) at 0x7f0f431ce7b0>,\n",
       "  <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431ceab0>,\n",
       "  <Feature CDS (chr1:1298370-1298421[-]) at 0x7f0f431ced80>,\n",
       "  <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431cf080>,\n",
       "  <Feature CDS (chr1:1298013-1298113[-]) at 0x7f0f431cf350>,\n",
       "  <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431cf650>,\n",
       "  <Feature CDS (chr1:1297822-1297933[-]) at 0x7f0f431cf920>,\n",
       "  <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431cfc20>,\n",
       "  <Feature CDS (chr1:1296425-1296633[-]) at 0x7f0f431cfef0>,\n",
       "  <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431e8230>,\n",
       "  <Feature CDS (chr1:1296211-1296280[-]) at 0x7f0f431e8500>,\n",
       "  <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431e8800>,\n",
       "  <Feature CDS (chr1:1296015-1296109[-]) at 0x7f0f431e8ad0>,\n",
       "  <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431e8dd0>,\n",
       "  <Feature CDS (chr1:1295736-1295938[-]) at 0x7f0f431e90a0>,\n",
       "  <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431e93a0>,\n",
       "  <Feature CDS (chr1:1295447-1295554[-]) at 0x7f0f431e9670>,\n",
       "  <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f431e9970>,\n",
       "  <Feature CDS (chr1:1294718-1294816[-]) at 0x7f0f431e9c40>,\n",
       "  <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431e9f40>,\n",
       "  <Feature CDS (chr1:1294402-1294628[-]) at 0x7f0f431ea210>,\n",
       "  <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431ea510>,\n",
       "  <Feature CDS (chr1:1294090-1294199[-]) at 0x7f0f431ea7e0>,\n",
       "  <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431eaae0>,\n",
       "  <Feature CDS (chr1:1293823-1293933[-]) at 0x7f0f431eadb0>,\n",
       "  <Feature exon (chr1:1292391-1293708[-]) at 0x7f0f431eb0b0>,\n",
       "  <Feature CDS (chr1:1293567-1293708[-]) at 0x7f0f431eb380>,\n",
       "  <Feature stop_codon (chr1:1293564-1293566[-]) at 0x7f0f431eb680>,\n",
       "  <Feature UTR (chr1:1307816-1307930[-]) at 0x7f0f431eb920>,\n",
       "  <Feature UTR (chr1:1292391-1293566[-]) at 0x7f0f431ebc20>,\n",
       "  <Feature transcript (chr1:1292870-1298955[-]) at 0x7f0f431ebef0>,\n",
       "  <Feature exon (chr1:1298567-1298955[-]) at 0x7f0f432001d0>,\n",
       "  <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f43200380>,\n",
       "  <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f43200560>,\n",
       "  <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f43200740>,\n",
       "  <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f43200920>,\n",
       "  <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f43200b00>,\n",
       "  <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f43200ce0>,\n",
       "  <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f43200ec0>,\n",
       "  <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f432010a0>,\n",
       "  <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f43201280>,\n",
       "  <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f43201460>,\n",
       "  <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f43201670>,\n",
       "  <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f43201880>,\n",
       "  <Feature exon (chr1:1292870-1293708[-]) at 0x7f0f43201a90>],\n",
       " [{'transcript_id': 'ENST00000416718',\n",
       "   'gene_name': 'MTCO3P12',\n",
       "   'transcript_type': 'unprocessed_pseudogene'},\n",
       "  {'transcript_id': 'ENST00000506640',\n",
       "   'gene_name': 'AL669831.1',\n",
       "   'transcript_type': 'lncRNA'},\n",
       "  {'transcript_id': 'ENST00000360001',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'protein_coding'},\n",
       "  {'transcript_id': 'ENST00000660930',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'protein_coding'},\n",
       "  {'transcript_id': 'ENST00000465727',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'nonsense_mediated_decay'},\n",
       "  {'transcript_id': 'ENST00000263741',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'protein_coding'},\n",
       "  {'transcript_id': 'ENST00000494748',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'retained_intron'},\n",
       "  {'transcript_id': 'ENST00000478938',\n",
       "   'gene_name': 'SDF4',\n",
       "   'transcript_type': 'retained_intron'},\n",
       "  {'transcript_id': 'ENST00000379198',\n",
       "   'gene_name': 'B3GALT6',\n",
       "   'transcript_type': 'protein_coding'},\n",
       "  {'transcript_id': 'ENST00000492936',\n",
       "   'gene_name': 'ACAP3',\n",
       "   'transcript_type': 'retained_intron'},\n",
       "  {'transcript_id': 'ENST00000354700',\n",
       "   'gene_name': 'ACAP3',\n",
       "   'transcript_type': 'protein_coding'},\n",
       "  {'transcript_id': 'ENST00000467278',\n",
       "   'gene_name': 'ACAP3',\n",
       "   'transcript_type': 'retained_intron'}])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract the unique transcript IDs from the modification sites\n",
    "transcript_ids_to_keep = set(modification_sites[\"transcript_id\"])\n",
    "\n",
    "# Filter features from the GTF database to include only relevant transcript IDs\n",
    "filtered_features = []\n",
    "transcript_data = []\n",
    "for feature in db.all_features():  # Iterate over all features in the database\n",
    "    if \"transcript_id\" in feature.attributes:\n",
    "        transcript_id = feature.attributes[\"transcript_id\"][0].split(\".\")[0]  # Remove version\n",
    "        if transcript_id in transcript_ids_to_keep:\n",
    "            filtered_features.append(feature)  # Keep only relevant features\n",
    "            if feature.featuretype == \"transcript\":\n",
    "                transcript_data.append({\n",
    "                \"transcript_id\": feature.id.split(\".\")[0],  # Remove version\n",
    "                \"gene_name\": feature.attributes.get(\"gene_name\", [None])[0],\n",
    "                \"transcript_type\": feature.attributes.get(\"transcript_type\", [None])[0],\n",
    "            })\n",
    "filtered_features, transcript_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1004da2a-2625-401c-a8fa-98ffbea99106",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id   gene_name          transcript_type\n",
       "0   ENST00000416718    MTCO3P12   unprocessed_pseudogene\n",
       "1   ENST00000506640  AL669831.1                   lncRNA\n",
       "2   ENST00000360001        SDF4           protein_coding\n",
       "3   ENST00000660930        SDF4           protein_coding\n",
       "4   ENST00000465727        SDF4  nonsense_mediated_decay\n",
       "5   ENST00000263741        SDF4           protein_coding\n",
       "6   ENST00000494748        SDF4          retained_intron\n",
       "7   ENST00000478938        SDF4          retained_intron\n",
       "8   ENST00000379198     B3GALT6           protein_coding\n",
       "9   ENST00000492936       ACAP3          retained_intron\n",
       "10  ENST00000354700       ACAP3           protein_coding\n",
       "11  ENST00000467278       ACAP3          retained_intron"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_data = pd.DataFrame(transcript_data).drop_duplicates(\"transcript_id\")\n",
    "transcript_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bc432e79-5f3e-47c2-832b-ca3bef60df32",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'ENST00000416718': [<Feature transcript (chr1:634376-634922[+]) at 0x7f0f43327950>,\n",
       "              <Feature exon (chr1:634376-634922[+]) at 0x7f0f43327e90>],\n",
       "             'ENST00000506640': [<Feature transcript (chr1:725885-778626[-]) at 0x7f0f4334cb90>,\n",
       "              <Feature exon (chr1:778284-778626[-]) at 0x7f0f4334ce30>,\n",
       "              <Feature exon (chr1:774171-774280[-]) at 0x7f0f4334c7d0>,\n",
       "              <Feature exon (chr1:772976-773107[-]) at 0x7f0f4334cd40>,\n",
       "              <Feature exon (chr1:769497-769712[-]) at 0x7f0f4334d0d0>,\n",
       "              <Feature exon (chr1:768548-768613[-]) at 0x7f0f4334d310>,\n",
       "              <Feature exon (chr1:766329-766387[-]) at 0x7f0f4334d520>,\n",
       "              <Feature exon (chr1:764723-765247[-]) at 0x7f0f4334d730>,\n",
       "              <Feature exon (chr1:758967-759123[-]) at 0x7f0f4334d940>,\n",
       "              <Feature exon (chr1:756077-756141[-]) at 0x7f0f4334db50>,\n",
       "              <Feature exon (chr1:746695-746818[-]) at 0x7f0f4334dd60>,\n",
       "              <Feature exon (chr1:744728-744825[-]) at 0x7f0f4334df70>,\n",
       "              <Feature exon (chr1:744195-744391[-]) at 0x7f0f4334e1b0>,\n",
       "              <Feature exon (chr1:743180-743350[-]) at 0x7f0f4334e3f0>,\n",
       "              <Feature exon (chr1:735423-735613[-]) at 0x7f0f4334e630>,\n",
       "              <Feature exon (chr1:733307-733364[-]) at 0x7f0f4334e870>,\n",
       "              <Feature exon (chr1:725885-729804[-]) at 0x7f0f4334eab0>],\n",
       "             'ENST00000360001': [<Feature transcript (chr1:1216908-1232001[-]) at 0x7f0f4334f1d0>,\n",
       "              <Feature exon (chr1:1231892-1232001[-]) at 0x7f0f4334f350>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f4334f1a0>,\n",
       "              <Feature CDS (chr1:1228468-1228772[-]) at 0x7f0f4334f0e0>,\n",
       "              <Feature start_codon (chr1:1228770-1228772[-]) at 0x7f0f4334f5c0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4334f7d0>,\n",
       "              <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f4334fa10>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4334fc80>,\n",
       "              <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f4334fec0>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43358170>,\n",
       "              <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f433583b0>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43358620>,\n",
       "              <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f43358860>,\n",
       "              <Feature exon (chr1:1216908-1217688[-]) at 0x7f0f43358ad0>,\n",
       "              <Feature CDS (chr1:1217515-1217688[-]) at 0x7f0f43358d10>,\n",
       "              <Feature stop_codon (chr1:1217512-1217514[-]) at 0x7f0f43358f80>,\n",
       "              <Feature UTR (chr1:1231892-1232001[-]) at 0x7f0f43359190>,\n",
       "              <Feature UTR (chr1:1228773-1228946[-]) at 0x7f0f43359400>,\n",
       "              <Feature UTR (chr1:1216908-1217514[-]) at 0x7f0f43359670>],\n",
       "             'ENST00000660930': [<Feature transcript (chr1:1216909-1232067[-]) at 0x7f0f433598e0>,\n",
       "              <Feature exon (chr1:1231892-1232067[-]) at 0x7f0f43359af0>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43359d60>,\n",
       "              <Feature CDS (chr1:1228468-1228793[-]) at 0x7f0f4335a000>,\n",
       "              <Feature start_codon (chr1:1228791-1228793[-]) at 0x7f0f4335a2d0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4335a540>,\n",
       "              <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f4335a7e0>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4335aab0>,\n",
       "              <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f4335ad50>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f4335b020>,\n",
       "              <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f4335b2c0>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f4335b590>,\n",
       "              <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f4335b830>,\n",
       "              <Feature exon (chr1:1216909-1217804[-]) at 0x7f0f4335bb00>,\n",
       "              <Feature CDS (chr1:1217673-1217804[-]) at 0x7f0f4335bda0>,\n",
       "              <Feature stop_codon (chr1:1217670-1217672[-]) at 0x7f0f433740b0>,\n",
       "              <Feature UTR (chr1:1231892-1232067[-]) at 0x7f0f43374320>,\n",
       "              <Feature UTR (chr1:1228794-1228946[-]) at 0x7f0f433745f0>,\n",
       "              <Feature UTR (chr1:1216909-1217672[-]) at 0x7f0f433748c0>],\n",
       "             'ENST00000465727': [<Feature transcript (chr1:1216931-1232031[-]) at 0x7f0f43374b90>,\n",
       "              <Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43374e00>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43374fb0>,\n",
       "              <Feature CDS (chr1:1228468-1228793[-]) at 0x7f0f43375190>,\n",
       "              <Feature start_codon (chr1:1228791-1228793[-]) at 0x7f0f433753a0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43375550>,\n",
       "              <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f43375730>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43375940>,\n",
       "              <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f43375b20>,\n",
       "              <Feature exon (chr1:1218769-1218943[-]) at 0x7f0f43375d30>,\n",
       "              <Feature CDS (chr1:1218769-1218943[-]) at 0x7f0f43375f10>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43376120>,\n",
       "              <Feature CDS (chr1:1218624-1218633[-]) at 0x7f0f43376300>,\n",
       "              <Feature stop_codon (chr1:1218621-1218623[-]) at 0x7f0f43376510>,\n",
       "              <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f433766c0>,\n",
       "              <Feature UTR (chr1:1231892-1232031[-]) at 0x7f0f433768a0>,\n",
       "              <Feature UTR (chr1:1228794-1228946[-]) at 0x7f0f43376ab0>,\n",
       "              <Feature UTR (chr1:1218458-1218623[-]) at 0x7f0f43376cc0>,\n",
       "              <Feature UTR (chr1:1216931-1217804[-]) at 0x7f0f43376ed0>],\n",
       "             'ENST00000263741': [<Feature transcript (chr1:1216931-1232031[-]) at 0x7f0f433770e0>,\n",
       "              <Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43377290>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f433774a0>,\n",
       "              <Feature CDS (chr1:1228468-1228772[-]) at 0x7f0f433776e0>,\n",
       "              <Feature start_codon (chr1:1228770-1228772[-]) at 0x7f0f43377950>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43377b60>,\n",
       "              <Feature CDS (chr1:1223832-1223968[-]) at 0x7f0f43377da0>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43377fe0>,\n",
       "              <Feature CDS (chr1:1223244-1223357[-]) at 0x7f0f43194290>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43194500>,\n",
       "              <Feature CDS (chr1:1218769-1218927[-]) at 0x7f0f43194740>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f431949b0>,\n",
       "              <Feature CDS (chr1:1218458-1218633[-]) at 0x7f0f43194bf0>,\n",
       "              <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f43194e60>,\n",
       "              <Feature CDS (chr1:1217673-1217804[-]) at 0x7f0f431950a0>,\n",
       "              <Feature stop_codon (chr1:1217670-1217672[-]) at 0x7f0f43195310>,\n",
       "              <Feature UTR (chr1:1231892-1232031[-]) at 0x7f0f43195520>,\n",
       "              <Feature UTR (chr1:1228773-1228946[-]) at 0x7f0f43195790>,\n",
       "              <Feature UTR (chr1:1216931-1217672[-]) at 0x7f0f43195a00>],\n",
       "             'ENST00000494748': [<Feature transcript (chr1:1216932-1221351[-]) at 0x7f0f43195c70>,\n",
       "              <Feature exon (chr1:1218769-1221351[-]) at 0x7f0f43195e80>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43196030>,\n",
       "              <Feature exon (chr1:1216932-1217688[-]) at 0x7f0f43196210>],\n",
       "             'ENST00000478938': [<Feature transcript (chr1:1216934-1220293[-]) at 0x7f0f431963f0>,\n",
       "              <Feature exon (chr1:1218458-1220293[-]) at 0x7f0f431965a0>,\n",
       "              <Feature exon (chr1:1216934-1217688[-]) at 0x7f0f43196750>],\n",
       "             'ENST00000379198': [<Feature transcript (chr1:1232237-1235041[+]) at 0x7f0f43196cf0>,\n",
       "              <Feature exon (chr1:1232237-1235041[+]) at 0x7f0f43196e10>,\n",
       "              <Feature CDS (chr1:1232279-1233265[+]) at 0x7f0f43196900>,\n",
       "              <Feature start_codon (chr1:1232279-1232281[+]) at 0x7f0f43197080>,\n",
       "              <Feature stop_codon (chr1:1233266-1233268[+]) at 0x7f0f43197320>,\n",
       "              <Feature UTR (chr1:1232237-1232278[+]) at 0x7f0f431975c0>,\n",
       "              <Feature UTR (chr1:1233266-1235041[+]) at 0x7f0f431978c0>],\n",
       "             'ENST00000492936': [<Feature transcript (chr1:1292390-1305929[-]) at 0x7f0f43197f80>,\n",
       "              <Feature exon (chr1:1304086-1305929[-]) at 0x7f0f43197ef0>,\n",
       "              <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f43197b90>,\n",
       "              <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b4290>,\n",
       "              <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431b4410>,\n",
       "              <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431b45f0>,\n",
       "              <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431b47d0>,\n",
       "              <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431b49b0>,\n",
       "              <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431b4b90>,\n",
       "              <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431b4d70>,\n",
       "              <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431b4f50>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431b5130>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431b5340>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431b5550>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431b5760>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431b5970>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431b5b80>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431b5d90>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431b5fa0>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431b61b0>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431b63c0>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431b65d0>,\n",
       "              <Feature exon (chr1:1292390-1293708[-]) at 0x7f0f431b67e0>],\n",
       "             'ENST00000354700': [<Feature transcript (chr1:1292391-1307930[-]) at 0x7f0f431b69f0>,\n",
       "              <Feature exon (chr1:1307769-1307930[-]) at 0x7f0f431b6bd0>,\n",
       "              <Feature CDS (chr1:1307769-1307815[-]) at 0x7f0f431b6e40>,\n",
       "              <Feature start_codon (chr1:1307813-1307815[-]) at 0x7f0f431b7110>,\n",
       "              <Feature exon (chr1:1304086-1304143[-]) at 0x7f0f431b7380>,\n",
       "              <Feature CDS (chr1:1304086-1304143[-]) at 0x7f0f431b7620>,\n",
       "              <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f431b78f0>,\n",
       "              <Feature CDS (chr1:1303162-1303281[-]) at 0x7f0f431b7b90>,\n",
       "              <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b7e60>,\n",
       "              <Feature CDS (chr1:1302922-1302975[-]) at 0x7f0f431cc140>,\n",
       "              <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431cc410>,\n",
       "              <Feature CDS (chr1:1301988-1302046[-]) at 0x7f0f431cc6b0>,\n",
       "              <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431cc980>,\n",
       "              <Feature CDS (chr1:1300509-1300692[-]) at 0x7f0f431ccc20>,\n",
       "              <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431ccef0>,\n",
       "              <Feature CDS (chr1:1300158-1300202[-]) at 0x7f0f431cd190>,\n",
       "              <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431cd460>,\n",
       "              <Feature CDS (chr1:1299973-1300068[-]) at 0x7f0f431cd700>,\n",
       "              <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431cd9d0>,\n",
       "              <Feature CDS (chr1:1299831-1299905[-]) at 0x7f0f431cdc70>,\n",
       "              <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431cdf40>,\n",
       "              <Feature CDS (chr1:1299345-1299356[-]) at 0x7f0f431ce1e0>,\n",
       "              <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431ce4e0>,\n",
       "              <Feature CDS (chr1:1298567-1298679[-]) at 0x7f0f431ce7b0>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431ceab0>,\n",
       "              <Feature CDS (chr1:1298370-1298421[-]) at 0x7f0f431ced80>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431cf080>,\n",
       "              <Feature CDS (chr1:1298013-1298113[-]) at 0x7f0f431cf350>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431cf650>,\n",
       "              <Feature CDS (chr1:1297822-1297933[-]) at 0x7f0f431cf920>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431cfc20>,\n",
       "              <Feature CDS (chr1:1296425-1296633[-]) at 0x7f0f431cfef0>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431e8230>,\n",
       "              <Feature CDS (chr1:1296211-1296280[-]) at 0x7f0f431e8500>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431e8800>,\n",
       "              <Feature CDS (chr1:1296015-1296109[-]) at 0x7f0f431e8ad0>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431e8dd0>,\n",
       "              <Feature CDS (chr1:1295736-1295938[-]) at 0x7f0f431e90a0>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431e93a0>,\n",
       "              <Feature CDS (chr1:1295447-1295554[-]) at 0x7f0f431e9670>,\n",
       "              <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f431e9970>,\n",
       "              <Feature CDS (chr1:1294718-1294816[-]) at 0x7f0f431e9c40>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431e9f40>,\n",
       "              <Feature CDS (chr1:1294402-1294628[-]) at 0x7f0f431ea210>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431ea510>,\n",
       "              <Feature CDS (chr1:1294090-1294199[-]) at 0x7f0f431ea7e0>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431eaae0>,\n",
       "              <Feature CDS (chr1:1293823-1293933[-]) at 0x7f0f431eadb0>,\n",
       "              <Feature exon (chr1:1292391-1293708[-]) at 0x7f0f431eb0b0>,\n",
       "              <Feature CDS (chr1:1293567-1293708[-]) at 0x7f0f431eb380>,\n",
       "              <Feature stop_codon (chr1:1293564-1293566[-]) at 0x7f0f431eb680>,\n",
       "              <Feature UTR (chr1:1307816-1307930[-]) at 0x7f0f431eb920>,\n",
       "              <Feature UTR (chr1:1292391-1293566[-]) at 0x7f0f431ebc20>],\n",
       "             'ENST00000467278': [<Feature transcript (chr1:1292870-1298955[-]) at 0x7f0f431ebef0>,\n",
       "              <Feature exon (chr1:1298567-1298955[-]) at 0x7f0f432001d0>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f43200380>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f43200560>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f43200740>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f43200920>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f43200b00>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f43200ce0>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f43200ec0>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f432010a0>,\n",
       "              <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f43201280>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f43201460>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f43201670>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f43201880>,\n",
       "              <Feature exon (chr1:1292870-1293708[-]) at 0x7f0f43201a90>]})"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group filtered features by transcript_id\n",
    "grouped_features = defaultdict(list)\n",
    "for feature in filtered_features:\n",
    "    transcript_id = feature.attributes[\"transcript_id\"][0].split(\".\")[0]  # Remove version\n",
    "    grouped_features[transcript_id].append(feature)\n",
    "grouped_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73434d73-b7b2-4481-8dee-2a85791d8e32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_length</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr5_length</th>\n",
       "      <th>utr3_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>6432</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>1956</td>\n",
       "      <td>1065</td>\n",
       "      <td>284</td>\n",
       "      <td>607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>2137</td>\n",
       "      <td>1044</td>\n",
       "      <td>329</td>\n",
       "      <td>764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>2095</td>\n",
       "      <td>762</td>\n",
       "      <td>293</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>2079</td>\n",
       "      <td>1023</td>\n",
       "      <td>314</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>3516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>2591</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>2805</td>\n",
       "      <td>987</td>\n",
       "      <td>42</td>\n",
       "      <td>1776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>5319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>3793</td>\n",
       "      <td>2502</td>\n",
       "      <td>115</td>\n",
       "      <td>1176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>2725</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  tx_length  cds_length  utr5_length  utr3_length\n",
       "0   ENST00000416718        547           0            0            0\n",
       "1   ENST00000506640       6432           0            0            0\n",
       "2   ENST00000360001       1956        1065          284          607\n",
       "3   ENST00000660930       2137        1044          329          764\n",
       "4   ENST00000465727       2095         762          293         1040\n",
       "5   ENST00000263741       2079        1023          314          742\n",
       "6   ENST00000494748       3516           0            0            0\n",
       "7   ENST00000478938       2591           0            0            0\n",
       "8   ENST00000379198       2805         987           42         1776\n",
       "9   ENST00000492936       5319           0            0            0\n",
       "10  ENST00000354700       3793        2502          115         1176\n",
       "11  ENST00000467278       2725           0            0            0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def calculate_transcript_features(grouped_features):\n",
    "    \"\"\"\n",
    "    Compute transcript features for grouped transcript data.\n",
    "    For each transcript:\n",
    "    - Sort CDS features by start and end positions to find start_codon and stop_codon positions.\n",
    "    - Calculate CDS length by summing all CDS feature lengths.\n",
    "    - Determine UTR5 and UTR3 lengths based on UTR positions relative to start_codon and stop_codon, considering the strand.\n",
    "    - If no CDS features exist, set CDS_len, utr5_len, and utr3_len to 0.\n",
    "    \"\"\"\n",
    "\n",
    "    transcript_features = []  # Store the results for all transcripts\n",
    "\n",
    "    for tx_id, features in grouped_features.items():\n",
    "        cds_length = 0\n",
    "        utr5_length = 0\n",
    "        utr3_length = 0\n",
    "        tx_length = 0\n",
    "        start_codon_pos = None\n",
    "        stop_codon_pos = None\n",
    "\n",
    "        # Separate features by type\n",
    "        cds_features = [f for f in features if f.featuretype == \"CDS\"]\n",
    "        utr_features = [f for f in features if f.featuretype == \"UTR\"]\n",
    "        exon_features = [f for f in features if f.featuretype == \"exon\"]\n",
    "        transcript_feature = next((f for f in features if f.featuretype == \"transcript\"), None)\n",
    "\n",
    "\n",
    "        # If no CDS features exist, set all lengths to 0\n",
    "        if not cds_features:\n",
    "            cds_length = 0\n",
    "            utr5_length = 0\n",
    "            utr3_length = 0\n",
    "            for exon in exon_features:\n",
    "                tx_length += exon.end - exon.start + 1\n",
    "\n",
    "        else:\n",
    "            # Sort CDS features by start and end positions\n",
    "            sorted_cds = sorted(cds_features, key=lambda x: (x.start, x.end))\n",
    "            if utr_features[0].strand == \"+\":  # Positive strand\n",
    "                start_codon_pos = sorted_cds[0].start  # First CDS start\n",
    "                stop_codon_pos = sorted_cds[-1].end  # Last CDS end\n",
    "            elif utr_features[0].strand == \"-\":  # Negative strand\n",
    "                start_codon_pos = sorted_cds[-1].end  # Last CDS end\n",
    "                stop_codon_pos = sorted_cds[0].start  # First CDS start\n",
    "\n",
    "            # Calculate CDS length by summing the lengths of all CDS features\n",
    "            for cds in sorted_cds:\n",
    "                cds_length += cds.end - cds.start + 1\n",
    "\n",
    "            # Handle UTRs and classify them as UTR5 or UTR3\n",
    "            if utr_features:\n",
    "                strand = utr_features[0].strand  # Assume all UTRs share the same strand\n",
    "                for utr in utr_features:\n",
    "                    utr_length = utr.end - utr.start + 1\n",
    "                    if strand == \"+\":\n",
    "                        if utr.end < start_codon_pos:  # 5' UTR is before start_codon\n",
    "                            utr5_length += utr_length\n",
    "                        elif utr.start > stop_codon_pos:  # 3' UTR is after stop_codon\n",
    "                            utr3_length += utr_length\n",
    "                    elif strand == \"-\":\n",
    "                        if utr.start > start_codon_pos:  # 5' UTR is after start_codon on negative strand\n",
    "                            utr5_length += utr_length\n",
    "                        elif utr.end < stop_codon_pos:  # 3' UTR is before stop_codon on negative strand\n",
    "                            utr3_length += utr_length\n",
    "            tx_length = cds_length + utr3_length + utr5_length\n",
    "       \n",
    "        # Add the results for this transcript\n",
    "        transcript_features.append({\n",
    "            \"transcript_id\": tx_id,\n",
    "            \"tx_length\": tx_length,\n",
    "            \"cds_length\": cds_length,\n",
    "            \"utr5_length\": utr5_length,\n",
    "            \"utr3_length\": utr3_length\n",
    "        })\n",
    "\n",
    "    return pd.DataFrame(transcript_features)\n",
    "# Compute transcript features for the filtered transcripts\n",
    "tx_features = calculate_transcript_features(grouped_features)\n",
    "tx_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "65af5bc2-03d8-42dc-935e-542f1b464fdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_length</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr5_length</th>\n",
       "      <th>utr3_length</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>6432</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>1956</td>\n",
       "      <td>1065</td>\n",
       "      <td>284</td>\n",
       "      <td>607</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>2137</td>\n",
       "      <td>1044</td>\n",
       "      <td>329</td>\n",
       "      <td>764</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>2095</td>\n",
       "      <td>762</td>\n",
       "      <td>293</td>\n",
       "      <td>1040</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>2079</td>\n",
       "      <td>1023</td>\n",
       "      <td>314</td>\n",
       "      <td>742</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>3516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>2591</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>2805</td>\n",
       "      <td>987</td>\n",
       "      <td>42</td>\n",
       "      <td>1776</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>5319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>3793</td>\n",
       "      <td>2502</td>\n",
       "      <td>115</td>\n",
       "      <td>1176</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>2725</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  tx_length  cds_length  utr5_length  utr3_length  \\\n",
       "0   ENST00000416718        547           0            0            0   \n",
       "1   ENST00000506640       6432           0            0            0   \n",
       "2   ENST00000360001       1956        1065          284          607   \n",
       "3   ENST00000660930       2137        1044          329          764   \n",
       "4   ENST00000465727       2095         762          293         1040   \n",
       "5   ENST00000263741       2079        1023          314          742   \n",
       "6   ENST00000494748       3516           0            0            0   \n",
       "7   ENST00000478938       2591           0            0            0   \n",
       "8   ENST00000379198       2805         987           42         1776   \n",
       "9   ENST00000492936       5319           0            0            0   \n",
       "10  ENST00000354700       3793        2502          115         1176   \n",
       "11  ENST00000467278       2725           0            0            0   \n",
       "\n",
       "     gene_name          transcript_type  \n",
       "0     MTCO3P12   unprocessed_pseudogene  \n",
       "1   AL669831.1                   lncRNA  \n",
       "2         SDF4           protein_coding  \n",
       "3         SDF4           protein_coding  \n",
       "4         SDF4  nonsense_mediated_decay  \n",
       "5         SDF4           protein_coding  \n",
       "6         SDF4          retained_intron  \n",
       "7         SDF4          retained_intron  \n",
       "8      B3GALT6           protein_coding  \n",
       "9        ACAP3          retained_intron  \n",
       "10       ACAP3           protein_coding  \n",
       "11       ACAP3          retained_intron  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_features = tx_features.merge(transcript_data, on=\"transcript_id\", how=\"left\")\n",
    "tx_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7075faeb-1d3d-4bb8-a32d-cca07ebe307b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'ENST00000416718': [<Feature exon (chr1:634376-634922[+]) at 0x7f0f43327e90>],\n",
       "             'ENST00000506640': [<Feature exon (chr1:778284-778626[-]) at 0x7f0f4334ce30>,\n",
       "              <Feature exon (chr1:774171-774280[-]) at 0x7f0f4334c7d0>,\n",
       "              <Feature exon (chr1:772976-773107[-]) at 0x7f0f4334cd40>,\n",
       "              <Feature exon (chr1:769497-769712[-]) at 0x7f0f4334d0d0>,\n",
       "              <Feature exon (chr1:768548-768613[-]) at 0x7f0f4334d310>,\n",
       "              <Feature exon (chr1:766329-766387[-]) at 0x7f0f4334d520>,\n",
       "              <Feature exon (chr1:764723-765247[-]) at 0x7f0f4334d730>,\n",
       "              <Feature exon (chr1:758967-759123[-]) at 0x7f0f4334d940>,\n",
       "              <Feature exon (chr1:756077-756141[-]) at 0x7f0f4334db50>,\n",
       "              <Feature exon (chr1:746695-746818[-]) at 0x7f0f4334dd60>,\n",
       "              <Feature exon (chr1:744728-744825[-]) at 0x7f0f4334df70>,\n",
       "              <Feature exon (chr1:744195-744391[-]) at 0x7f0f4334e1b0>,\n",
       "              <Feature exon (chr1:743180-743350[-]) at 0x7f0f4334e3f0>,\n",
       "              <Feature exon (chr1:735423-735613[-]) at 0x7f0f4334e630>,\n",
       "              <Feature exon (chr1:733307-733364[-]) at 0x7f0f4334e870>,\n",
       "              <Feature exon (chr1:725885-729804[-]) at 0x7f0f4334eab0>],\n",
       "             'ENST00000360001': [<Feature exon (chr1:1231892-1232001[-]) at 0x7f0f4334f350>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f4334f1a0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4334f7d0>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4334fc80>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43358170>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43358620>,\n",
       "              <Feature exon (chr1:1216908-1217688[-]) at 0x7f0f43358ad0>],\n",
       "             'ENST00000660930': [<Feature exon (chr1:1231892-1232067[-]) at 0x7f0f43359af0>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43359d60>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f4335a540>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f4335aab0>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f4335b020>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f4335b590>,\n",
       "              <Feature exon (chr1:1216909-1217804[-]) at 0x7f0f4335bb00>],\n",
       "             'ENST00000465727': [<Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43374e00>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f43374fb0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43375550>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43375940>,\n",
       "              <Feature exon (chr1:1218769-1218943[-]) at 0x7f0f43375d30>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43376120>,\n",
       "              <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f433766c0>],\n",
       "             'ENST00000263741': [<Feature exon (chr1:1231892-1232031[-]) at 0x7f0f43377290>,\n",
       "              <Feature exon (chr1:1228468-1228946[-]) at 0x7f0f433774a0>,\n",
       "              <Feature exon (chr1:1223832-1223968[-]) at 0x7f0f43377b60>,\n",
       "              <Feature exon (chr1:1223244-1223357[-]) at 0x7f0f43377fe0>,\n",
       "              <Feature exon (chr1:1218769-1218927[-]) at 0x7f0f43194500>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f431949b0>,\n",
       "              <Feature exon (chr1:1216931-1217804[-]) at 0x7f0f43194e60>],\n",
       "             'ENST00000494748': [<Feature exon (chr1:1218769-1221351[-]) at 0x7f0f43195e80>,\n",
       "              <Feature exon (chr1:1218458-1218633[-]) at 0x7f0f43196030>,\n",
       "              <Feature exon (chr1:1216932-1217688[-]) at 0x7f0f43196210>],\n",
       "             'ENST00000478938': [<Feature exon (chr1:1218458-1220293[-]) at 0x7f0f431965a0>,\n",
       "              <Feature exon (chr1:1216934-1217688[-]) at 0x7f0f43196750>],\n",
       "             'ENST00000379198': [<Feature exon (chr1:1232237-1235041[+]) at 0x7f0f43196e10>],\n",
       "             'ENST00000492936': [<Feature exon (chr1:1304086-1305929[-]) at 0x7f0f43197ef0>,\n",
       "              <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f43197b90>,\n",
       "              <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b4290>,\n",
       "              <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431b4410>,\n",
       "              <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431b45f0>,\n",
       "              <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431b47d0>,\n",
       "              <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431b49b0>,\n",
       "              <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431b4b90>,\n",
       "              <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431b4d70>,\n",
       "              <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431b4f50>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431b5130>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431b5340>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431b5550>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431b5760>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431b5970>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431b5b80>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431b5d90>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431b5fa0>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431b61b0>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431b63c0>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431b65d0>,\n",
       "              <Feature exon (chr1:1292390-1293708[-]) at 0x7f0f431b67e0>],\n",
       "             'ENST00000354700': [<Feature exon (chr1:1307769-1307930[-]) at 0x7f0f431b6bd0>,\n",
       "              <Feature exon (chr1:1304086-1304143[-]) at 0x7f0f431b7380>,\n",
       "              <Feature exon (chr1:1303162-1303281[-]) at 0x7f0f431b78f0>,\n",
       "              <Feature exon (chr1:1302922-1302975[-]) at 0x7f0f431b7e60>,\n",
       "              <Feature exon (chr1:1301988-1302046[-]) at 0x7f0f431cc410>,\n",
       "              <Feature exon (chr1:1300509-1300692[-]) at 0x7f0f431cc980>,\n",
       "              <Feature exon (chr1:1300158-1300202[-]) at 0x7f0f431ccef0>,\n",
       "              <Feature exon (chr1:1299973-1300068[-]) at 0x7f0f431cd460>,\n",
       "              <Feature exon (chr1:1299831-1299905[-]) at 0x7f0f431cd9d0>,\n",
       "              <Feature exon (chr1:1299345-1299356[-]) at 0x7f0f431cdf40>,\n",
       "              <Feature exon (chr1:1298567-1298679[-]) at 0x7f0f431ce4e0>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f431ceab0>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f431cf080>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f431cf650>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f431cfc20>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f431e8230>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f431e8800>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f431e8dd0>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f431e93a0>,\n",
       "              <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f431e9970>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f431e9f40>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f431ea510>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f431eaae0>,\n",
       "              <Feature exon (chr1:1292391-1293708[-]) at 0x7f0f431eb0b0>],\n",
       "             'ENST00000467278': [<Feature exon (chr1:1298567-1298955[-]) at 0x7f0f432001d0>,\n",
       "              <Feature exon (chr1:1298370-1298421[-]) at 0x7f0f43200380>,\n",
       "              <Feature exon (chr1:1298013-1298113[-]) at 0x7f0f43200560>,\n",
       "              <Feature exon (chr1:1297822-1297933[-]) at 0x7f0f43200740>,\n",
       "              <Feature exon (chr1:1296425-1296633[-]) at 0x7f0f43200920>,\n",
       "              <Feature exon (chr1:1296211-1296280[-]) at 0x7f0f43200b00>,\n",
       "              <Feature exon (chr1:1296015-1296109[-]) at 0x7f0f43200ce0>,\n",
       "              <Feature exon (chr1:1295736-1295938[-]) at 0x7f0f43200ec0>,\n",
       "              <Feature exon (chr1:1295447-1295554[-]) at 0x7f0f432010a0>,\n",
       "              <Feature exon (chr1:1294718-1294816[-]) at 0x7f0f43201280>,\n",
       "              <Feature exon (chr1:1294402-1294628[-]) at 0x7f0f43201460>,\n",
       "              <Feature exon (chr1:1294090-1294199[-]) at 0x7f0f43201670>,\n",
       "              <Feature exon (chr1:1293823-1293933[-]) at 0x7f0f43201880>,\n",
       "              <Feature exon (chr1:1292870-1293708[-]) at 0x7f0f43201a90>]})"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter exons for relevant transcripts\n",
    "exons = defaultdict(list)\n",
    "for tx_id, features in grouped_features.items():\n",
    "    for feature in features:\n",
    "        if feature.featuretype == \"exon\":\n",
    "            exons[tx_id].append(feature)\n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "64ba333e-f3d3-43ff-b18d-0fe0c1ca571b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert transcript coordinates to genomic coordinates\n",
    "def convert_to_genome_coordinates(tx_name, tx_pos, exonsdb, txfdb):\n",
    "    \"\"\"\n",
    "    Convert transcript coordinates to genome coordinates and annotate regions.\n",
    "    \"\"\"\n",
    "    if tx_name not in exonsdb:\n",
    "        warnings.warn(f\"Transcript {tx_name} not found in exons database.\")\n",
    "        return {\n",
    "            \"genome_pos\": None,\n",
    "            \"chromosome\": None,\n",
    "            \"dist_up_exon_junc\": None,\n",
    "            \"dist_down_exon_junc\": None,\n",
    "            \"region\": None\n",
    "        }\n",
    "    \n",
    "    tx_exons = sorted(exonsdb[tx_name], key=lambda x: x.start)  # Sort exons by start\n",
    "    chrom = tx_exons[0].chrom\n",
    "    strand = tx_exons[0].strand\n",
    "    \n",
    "    # Calculate cumulative exon lengths\n",
    "    exon_lengths = [exon.end - exon.start + 1 for exon in tx_exons]\n",
    "    cum_lengths = np.cumsum(exon_lengths)\n",
    "\n",
    "    # Identify the exon containing the position\n",
    "    idx = np.searchsorted(cum_lengths, tx_pos)\n",
    "    \n",
    "    if idx >= len(tx_exons):\n",
    "        warnings.warn(f\"Transcript position {tx_pos} is out of range for transcript {tx_name}\")\n",
    "        return {\n",
    "            \"genome_pos\": None,\n",
    "            \"chromosome\": None,\n",
    "            \"dist_up_exon_junc\": None,\n",
    "            \"dist_down_exon_junc\": None,\n",
    "            \"region\": None\n",
    "        }\n",
    "    \n",
    "    # Compute genomic position\n",
    "    rel_pos = tx_pos - (cum_lengths[idx - 1] if idx > 0 else 0)\n",
    "    exon = tx_exons[idx]\n",
    "    if strand == \"+\":\n",
    "        genome_pos = exon.start + rel_pos - 1\n",
    "    else:\n",
    "        genome_pos = exon.end - rel_pos + 1\n",
    "    \n",
    "    # Determine distance to exon junctions\n",
    "    dist_upstream = rel_pos\n",
    "    dist_downstream = exon.end - exon.start - rel_pos\n",
    "\n",
    "    # Determine region (UTR, CDS, etc.)\n",
    "    tx_feat = txfdb[txfdb[\"transcript_id\"] == tx_name].iloc[0]\n",
    "    if tx_feat[\"cds_length\"] == 0:\n",
    "        region = \"ncRNA\"\n",
    "    elif tx_pos <= tx_feat[\"utr5_length\"]:\n",
    "        region = \"UTR5\"\n",
    "    elif tx_pos <= tx_feat[\"utr5_length\"] + tx_feat[\"cds_length\"]:\n",
    "        region = \"CDS\"\n",
    "    elif tx_pos <= tx_feat[\"tx_length\"]:\n",
    "        region = \"UTR3\"\n",
    "    else:\n",
    "        region = None\n",
    "    \n",
    "    return {\n",
    "        \"genome_pos\": genome_pos,\n",
    "        \"chromosome\": chrom,\n",
    "        \"dist_up_exon_junc\": dist_upstream,\n",
    "        \"dist_down_exon_junc\": dist_downstream,\n",
    "        \"region\": region\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e45dbe12-cd28-4c9c-9592-99694320a39b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 774177, 'chromosome': 'chr1', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 778545, 'chromosome': 'chr1', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 634716, 'chromosome': 'chr1', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1223879, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1228561, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1228927, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1228941, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1228941, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1219381, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1218948, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1219068, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1218635, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1234668, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1234815, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1234848, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1304658, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1300627, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>{'genome_pos': 1298659, 'chromosome': 'chr1', ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  \\\n",
       "0               0.980850  GAACT   0.655172        s1     caud   \n",
       "1               0.965541  AGACT   0.913043        s1     caud   \n",
       "2               0.927230  GGACA   0.666667        s1     caud   \n",
       "3               0.936434  GGACT   0.687898        s1     caud   \n",
       "4               0.938970  AAACT   0.567010        s1     caud   \n",
       "5               0.969710  GGACT   0.777778        s2     caud   \n",
       "6               0.949604  GGACT   0.761905        s2     caud   \n",
       "7               0.928761  GGACT   0.761905        s2     caud   \n",
       "8               0.941420  GGACT   0.765432        s2     caud   \n",
       "9               0.903208  AAACT   0.558559        s2     caud   \n",
       "10              0.936208  GGACT   0.779412        s2     caud   \n",
       "11              0.911677  AAACT   0.556701        s2     caud   \n",
       "12              0.987061  GGACT   0.909091        s2     caud   \n",
       "13              0.995830  GGACA   0.962963        s2     caud   \n",
       "14              0.952746  GGACT   0.677419        s2     caud   \n",
       "15              0.956273  AGACA   0.770492        s2     caud   \n",
       "16              0.955948  AGACA   0.794872        s2     caud   \n",
       "17              0.966780  AGACA   0.868421        s2     caud   \n",
       "\n",
       "                                           annotation  \n",
       "0   {'genome_pos': 774177, 'chromosome': 'chr1', '...  \n",
       "1   {'genome_pos': 778545, 'chromosome': 'chr1', '...  \n",
       "2   {'genome_pos': 634716, 'chromosome': 'chr1', '...  \n",
       "3   {'genome_pos': 1223879, 'chromosome': 'chr1', ...  \n",
       "4   {'genome_pos': 1228561, 'chromosome': 'chr1', ...  \n",
       "5   {'genome_pos': 1228927, 'chromosome': 'chr1', ...  \n",
       "6   {'genome_pos': 1228941, 'chromosome': 'chr1', ...  \n",
       "7   {'genome_pos': 1228941, 'chromosome': 'chr1', ...  \n",
       "8   {'genome_pos': 1219381, 'chromosome': 'chr1', ...  \n",
       "9   {'genome_pos': 1218948, 'chromosome': 'chr1', ...  \n",
       "10  {'genome_pos': 1219068, 'chromosome': 'chr1', ...  \n",
       "11  {'genome_pos': 1218635, 'chromosome': 'chr1', ...  \n",
       "12  {'genome_pos': 1234668, 'chromosome': 'chr1', ...  \n",
       "13  {'genome_pos': 1234815, 'chromosome': 'chr1', ...  \n",
       "14  {'genome_pos': 1234848, 'chromosome': 'chr1', ...  \n",
       "15  {'genome_pos': 1304658, 'chromosome': 'chr1', ...  \n",
       "16  {'genome_pos': 1300627, 'chromosome': 'chr1', ...  \n",
       "17  {'genome_pos': 1298659, 'chromosome': 'chr1', ...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modification_sites[\"annotation\"] = modification_sites.apply(\n",
    "    lambda row: convert_to_genome_coordinates(row[\"transcript_id\"], row[\"transcript_position\"],exons, tx_features), axis=1\n",
    ")\n",
    "modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "12913ee2-a15f-4e0f-a65d-e7aa3ca90802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>genome_pos</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dist_up_exon_junc</th>\n",
       "      <th>dist_down_exon_junc</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>774177</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>778545</td>\n",
       "      <td>chr1</td>\n",
       "      <td>82</td>\n",
       "      <td>260</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>634716</td>\n",
       "      <td>chr1</td>\n",
       "      <td>341</td>\n",
       "      <td>205</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1223879</td>\n",
       "      <td>chr1</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228561</td>\n",
       "      <td>chr1</td>\n",
       "      <td>386</td>\n",
       "      <td>92</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>458</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219381</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1971</td>\n",
       "      <td>611</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218948</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2404</td>\n",
       "      <td>178</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219068</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1226</td>\n",
       "      <td>609</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1659</td>\n",
       "      <td>176</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234668</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2432</td>\n",
       "      <td>372</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234815</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2579</td>\n",
       "      <td>225</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234848</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2612</td>\n",
       "      <td>192</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1304658</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1272</td>\n",
       "      <td>571</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1300627</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1298659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>297</td>\n",
       "      <td>91</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  genome_pos  \\\n",
       "0               0.980850  GAACT   0.655172        s1     caud      774177   \n",
       "1               0.965541  AGACT   0.913043        s1     caud      778545   \n",
       "2               0.927230  GGACA   0.666667        s1     caud      634716   \n",
       "3               0.936434  GGACT   0.687898        s1     caud     1223879   \n",
       "4               0.938970  AAACT   0.567010        s1     caud     1228561   \n",
       "5               0.969710  GGACT   0.777778        s2     caud     1228927   \n",
       "6               0.949604  GGACT   0.761905        s2     caud     1228941   \n",
       "7               0.928761  GGACT   0.761905        s2     caud     1228941   \n",
       "8               0.941420  GGACT   0.765432        s2     caud     1219381   \n",
       "9               0.903208  AAACT   0.558559        s2     caud     1218948   \n",
       "10              0.936208  GGACT   0.779412        s2     caud     1219068   \n",
       "11              0.911677  AAACT   0.556701        s2     caud     1218635   \n",
       "12              0.987061  GGACT   0.909091        s2     caud     1234668   \n",
       "13              0.995830  GGACA   0.962963        s2     caud     1234815   \n",
       "14              0.952746  GGACT   0.677419        s2     caud     1234848   \n",
       "15              0.956273  AGACA   0.770492        s2     caud     1304658   \n",
       "16              0.955948  AGACA   0.794872        s2     caud     1300627   \n",
       "17              0.966780  AGACA   0.868421        s2     caud     1298659   \n",
       "\n",
       "   chromosome  dist_up_exon_junc  dist_down_exon_junc region  \n",
       "0        chr1                104                    5  ncRNA  \n",
       "1        chr1                 82                  260  ncRNA  \n",
       "2        chr1                341                  205  ncRNA  \n",
       "3        chr1                 90                   46    CDS  \n",
       "4        chr1                386                   92   UTR3  \n",
       "5        chr1                 20                  458   UTR3  \n",
       "6        chr1                  6                  472   UTR3  \n",
       "7        chr1                  6                  472   UTR3  \n",
       "8        chr1               1971                  611  ncRNA  \n",
       "9        chr1               2404                  178  ncRNA  \n",
       "10       chr1               1226                  609  ncRNA  \n",
       "11       chr1               1659                  176  ncRNA  \n",
       "12       chr1               2432                  372   UTR3  \n",
       "13       chr1               2579                  225   UTR3  \n",
       "14       chr1               2612                  192   UTR3  \n",
       "15       chr1               1272                  571  ncRNA  \n",
       "16       chr1                 66                  117   UTR3  \n",
       "17       chr1                297                   91  ncRNA  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Expand dictionary columns\n",
    "annotation_df = modification_sites[\"annotation\"].apply(pd.Series)\n",
    "modification_sites = pd.concat([modification_sites, annotation_df], axis=1).drop(columns=[\"annotation\"])\n",
    "modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "fd04221d-a8fb-42c6-9d07-d536a8a6eeb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>genome_pos</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dist_up_exon_junc</th>\n",
       "      <th>dist_down_exon_junc</th>\n",
       "      <th>region</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>774177</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>778545</td>\n",
       "      <td>chr1</td>\n",
       "      <td>82</td>\n",
       "      <td>260</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>634716</td>\n",
       "      <td>chr1</td>\n",
       "      <td>341</td>\n",
       "      <td>205</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1223879</td>\n",
       "      <td>chr1</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228561</td>\n",
       "      <td>chr1</td>\n",
       "      <td>386</td>\n",
       "      <td>92</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>458</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219381</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1971</td>\n",
       "      <td>611</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218948</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2404</td>\n",
       "      <td>178</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219068</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1226</td>\n",
       "      <td>609</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1659</td>\n",
       "      <td>176</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234668</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2432</td>\n",
       "      <td>372</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234815</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2579</td>\n",
       "      <td>225</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234848</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2612</td>\n",
       "      <td>192</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1304658</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1272</td>\n",
       "      <td>571</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1300627</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1298659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>297</td>\n",
       "      <td>91</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  genome_pos  \\\n",
       "0               0.980850  GAACT   0.655172        s1     caud      774177   \n",
       "1               0.965541  AGACT   0.913043        s1     caud      778545   \n",
       "2               0.927230  GGACA   0.666667        s1     caud      634716   \n",
       "3               0.936434  GGACT   0.687898        s1     caud     1223879   \n",
       "4               0.938970  AAACT   0.567010        s1     caud     1228561   \n",
       "5               0.969710  GGACT   0.777778        s2     caud     1228927   \n",
       "6               0.949604  GGACT   0.761905        s2     caud     1228941   \n",
       "7               0.928761  GGACT   0.761905        s2     caud     1228941   \n",
       "8               0.941420  GGACT   0.765432        s2     caud     1219381   \n",
       "9               0.903208  AAACT   0.558559        s2     caud     1218948   \n",
       "10              0.936208  GGACT   0.779412        s2     caud     1219068   \n",
       "11              0.911677  AAACT   0.556701        s2     caud     1218635   \n",
       "12              0.987061  GGACT   0.909091        s2     caud     1234668   \n",
       "13              0.995830  GGACA   0.962963        s2     caud     1234815   \n",
       "14              0.952746  GGACT   0.677419        s2     caud     1234848   \n",
       "15              0.956273  AGACA   0.770492        s2     caud     1304658   \n",
       "16              0.955948  AGACA   0.794872        s2     caud     1300627   \n",
       "17              0.966780  AGACA   0.868421        s2     caud     1298659   \n",
       "\n",
       "   chromosome  dist_up_exon_junc  dist_down_exon_junc region   gene_name  \\\n",
       "0        chr1                104                    5  ncRNA  AL669831.1   \n",
       "1        chr1                 82                  260  ncRNA  AL669831.1   \n",
       "2        chr1                341                  205  ncRNA    MTCO3P12   \n",
       "3        chr1                 90                   46    CDS        SDF4   \n",
       "4        chr1                386                   92   UTR3        SDF4   \n",
       "5        chr1                 20                  458   UTR3        SDF4   \n",
       "6        chr1                  6                  472   UTR3        SDF4   \n",
       "7        chr1                  6                  472   UTR3        SDF4   \n",
       "8        chr1               1971                  611  ncRNA        SDF4   \n",
       "9        chr1               2404                  178  ncRNA        SDF4   \n",
       "10       chr1               1226                  609  ncRNA        SDF4   \n",
       "11       chr1               1659                  176  ncRNA        SDF4   \n",
       "12       chr1               2432                  372   UTR3     B3GALT6   \n",
       "13       chr1               2579                  225   UTR3     B3GALT6   \n",
       "14       chr1               2612                  192   UTR3     B3GALT6   \n",
       "15       chr1               1272                  571  ncRNA       ACAP3   \n",
       "16       chr1                 66                  117   UTR3       ACAP3   \n",
       "17       chr1                297                   91  ncRNA       ACAP3   \n",
       "\n",
       "            transcript_type  \n",
       "0                    lncRNA  \n",
       "1                    lncRNA  \n",
       "2    unprocessed_pseudogene  \n",
       "3            protein_coding  \n",
       "4            protein_coding  \n",
       "5            protein_coding  \n",
       "6   nonsense_mediated_decay  \n",
       "7            protein_coding  \n",
       "8           retained_intron  \n",
       "9           retained_intron  \n",
       "10          retained_intron  \n",
       "11          retained_intron  \n",
       "12           protein_coding  \n",
       "13           protein_coding  \n",
       "14           protein_coding  \n",
       "15          retained_intron  \n",
       "16           protein_coding  \n",
       "17          retained_intron  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with transcript biotype information\n",
    "annotated_modification_sites = modification_sites.merge(transcript_data, on=\"transcript_id\", how=\"left\")\n",
    "annotated_modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "479bcea2-f7b5-4fd5-a18d-2ce4abb0a9ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>genome_pos</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dist_up_exon_junc</th>\n",
       "      <th>dist_down_exon_junc</th>\n",
       "      <th>region</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1300627</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1223879</td>\n",
       "      <td>chr1</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228561</td>\n",
       "      <td>chr1</td>\n",
       "      <td>386</td>\n",
       "      <td>92</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234848</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2612</td>\n",
       "      <td>192</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234815</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2579</td>\n",
       "      <td>225</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234668</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2432</td>\n",
       "      <td>372</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>634716</td>\n",
       "      <td>chr1</td>\n",
       "      <td>341</td>\n",
       "      <td>205</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MTCO3P12</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1298659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>297</td>\n",
       "      <td>91</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219068</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1226</td>\n",
       "      <td>609</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1659</td>\n",
       "      <td>176</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1304658</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1272</td>\n",
       "      <td>571</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219381</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1971</td>\n",
       "      <td>611</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218948</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2404</td>\n",
       "      <td>178</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>778545</td>\n",
       "      <td>chr1</td>\n",
       "      <td>82</td>\n",
       "      <td>260</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>774177</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>458</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  genome_pos  \\\n",
       "7               0.928761  GGACT   0.761905        s2     caud     1228941   \n",
       "16              0.955948  AGACA   0.794872        s2     caud     1300627   \n",
       "3               0.936434  GGACT   0.687898        s1     caud     1223879   \n",
       "4               0.938970  AAACT   0.567010        s1     caud     1228561   \n",
       "14              0.952746  GGACT   0.677419        s2     caud     1234848   \n",
       "13              0.995830  GGACA   0.962963        s2     caud     1234815   \n",
       "12              0.987061  GGACT   0.909091        s2     caud     1234668   \n",
       "2               0.927230  GGACA   0.666667        s1     caud      634716   \n",
       "6               0.949604  GGACT   0.761905        s2     caud     1228941   \n",
       "17              0.966780  AGACA   0.868421        s2     caud     1298659   \n",
       "10              0.936208  GGACT   0.779412        s2     caud     1219068   \n",
       "11              0.911677  AAACT   0.556701        s2     caud     1218635   \n",
       "15              0.956273  AGACA   0.770492        s2     caud     1304658   \n",
       "8               0.941420  GGACT   0.765432        s2     caud     1219381   \n",
       "9               0.903208  AAACT   0.558559        s2     caud     1218948   \n",
       "1               0.965541  AGACT   0.913043        s1     caud      778545   \n",
       "0               0.980850  GAACT   0.655172        s1     caud      774177   \n",
       "5               0.969710  GGACT   0.777778        s2     caud     1228927   \n",
       "\n",
       "   chromosome  dist_up_exon_junc  dist_down_exon_junc region   gene_name  \\\n",
       "7        chr1                  6                  472   UTR3        SDF4   \n",
       "16       chr1                 66                  117   UTR3       ACAP3   \n",
       "3        chr1                 90                   46    CDS        SDF4   \n",
       "4        chr1                386                   92   UTR3        SDF4   \n",
       "14       chr1               2612                  192   UTR3     B3GALT6   \n",
       "13       chr1               2579                  225   UTR3     B3GALT6   \n",
       "12       chr1               2432                  372   UTR3     B3GALT6   \n",
       "2        chr1                341                  205  ncRNA    MTCO3P12   \n",
       "6        chr1                  6                  472   UTR3        SDF4   \n",
       "17       chr1                297                   91  ncRNA       ACAP3   \n",
       "10       chr1               1226                  609  ncRNA        SDF4   \n",
       "11       chr1               1659                  176  ncRNA        SDF4   \n",
       "15       chr1               1272                  571  ncRNA       ACAP3   \n",
       "8        chr1               1971                  611  ncRNA        SDF4   \n",
       "9        chr1               2404                  178  ncRNA        SDF4   \n",
       "1        chr1                 82                  260  ncRNA  AL669831.1   \n",
       "0        chr1                104                    5  ncRNA  AL669831.1   \n",
       "5        chr1                 20                  458   UTR3        SDF4   \n",
       "\n",
       "            transcript_type  \n",
       "7            protein_coding  \n",
       "16           protein_coding  \n",
       "3            protein_coding  \n",
       "4            protein_coding  \n",
       "14           protein_coding  \n",
       "13           protein_coding  \n",
       "12           protein_coding  \n",
       "2    unprocessed_pseudogene  \n",
       "6   nonsense_mediated_decay  \n",
       "17          retained_intron  \n",
       "10          retained_intron  \n",
       "11          retained_intron  \n",
       "15          retained_intron  \n",
       "8           retained_intron  \n",
       "9           retained_intron  \n",
       "1                    lncRNA  \n",
       "0                    lncRNA  \n",
       "5            protein_coding  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_modification_sites = annotated_modification_sites.sort_values(by='transcript_id')\n",
    "annotated_modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d9987b75-2acd-4bf2-89ac-6d392c300889",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotated modification sites saved to 4_test_output_test_data_m6anet_mod_sites.csv\n"
     ]
    }
   ],
   "source": [
    "# Save output\n",
    "output_prefix = 'test_output'\n",
    "output_file = f\"4_{output_prefix}_{f}\"\n",
    "annotated_modification_sites.to_csv(output_file, index=False)\n",
    "\n",
    "print(\"Annotated modification sites saved to\", output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "063a20da-5350-4671-950e-4d3b5f6cd72b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>genome_pos</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dist_up_exon_junc</th>\n",
       "      <th>dist_down_exon_junc</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>774177</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>778545</td>\n",
       "      <td>chr1</td>\n",
       "      <td>82</td>\n",
       "      <td>260</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>634716</td>\n",
       "      <td>chr1</td>\n",
       "      <td>341</td>\n",
       "      <td>205</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1223879</td>\n",
       "      <td>chr1</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228561</td>\n",
       "      <td>chr1</td>\n",
       "      <td>386</td>\n",
       "      <td>92</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>458</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219381</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1971</td>\n",
       "      <td>611</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218948</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2404</td>\n",
       "      <td>178</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219068</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1226</td>\n",
       "      <td>609</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1659</td>\n",
       "      <td>176</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234668</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2432</td>\n",
       "      <td>372</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234815</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2579</td>\n",
       "      <td>225</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234848</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2612</td>\n",
       "      <td>192</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1304658</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1272</td>\n",
       "      <td>571</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1300627</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1298659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>297</td>\n",
       "      <td>91</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  genome_pos  \\\n",
       "0               0.980850  GAACT   0.655172        s1     caud      774177   \n",
       "1               0.965541  AGACT   0.913043        s1     caud      778545   \n",
       "2               0.927230  GGACA   0.666667        s1     caud      634716   \n",
       "3               0.936434  GGACT   0.687898        s1     caud     1223879   \n",
       "4               0.938970  AAACT   0.567010        s1     caud     1228561   \n",
       "5               0.969710  GGACT   0.777778        s2     caud     1228927   \n",
       "6               0.949604  GGACT   0.761905        s2     caud     1228941   \n",
       "7               0.928761  GGACT   0.761905        s2     caud     1228941   \n",
       "8               0.941420  GGACT   0.765432        s2     caud     1219381   \n",
       "9               0.903208  AAACT   0.558559        s2     caud     1218948   \n",
       "10              0.936208  GGACT   0.779412        s2     caud     1219068   \n",
       "11              0.911677  AAACT   0.556701        s2     caud     1218635   \n",
       "12              0.987061  GGACT   0.909091        s2     caud     1234668   \n",
       "13              0.995830  GGACA   0.962963        s2     caud     1234815   \n",
       "14              0.952746  GGACT   0.677419        s2     caud     1234848   \n",
       "15              0.956273  AGACA   0.770492        s2     caud     1304658   \n",
       "16              0.955948  AGACA   0.794872        s2     caud     1300627   \n",
       "17              0.966780  AGACA   0.868421        s2     caud     1298659   \n",
       "\n",
       "   chromosome  dist_up_exon_junc  dist_down_exon_junc region  \n",
       "0        chr1                104                    5  ncRNA  \n",
       "1        chr1                 82                  260  ncRNA  \n",
       "2        chr1                341                  205  ncRNA  \n",
       "3        chr1                 90                   46    CDS  \n",
       "4        chr1                386                   92   UTR3  \n",
       "5        chr1                 20                  458   UTR3  \n",
       "6        chr1                  6                  472   UTR3  \n",
       "7        chr1                  6                  472   UTR3  \n",
       "8        chr1               1971                  611  ncRNA  \n",
       "9        chr1               2404                  178  ncRNA  \n",
       "10       chr1               1226                  609  ncRNA  \n",
       "11       chr1               1659                  176  ncRNA  \n",
       "12       chr1               2432                  372   UTR3  \n",
       "13       chr1               2579                  225   UTR3  \n",
       "14       chr1               2612                  192   UTR3  \n",
       "15       chr1               1272                  571  ncRNA  \n",
       "16       chr1                 66                  117   UTR3  \n",
       "17       chr1                297                   91  ncRNA  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for _, row in modification_sites.iterrows():  # Iterate over rows in the DataFrame\n",
    "    # Process the row to compute genomic coordinates\n",
    "    result = convert_to_genome_coordinates(\n",
    "        row[\"transcript_id\"], row[\"transcript_position\"], exons, tx_features\n",
    "    )\n",
    "\n",
    "    # Merge the original row data into the result dictionary\n",
    "    result.update(row.to_dict())\n",
    "\n",
    "    # Reorder the fields in the desired order\n",
    "    reordered_result = {\n",
    "        # Fields from the original modification_sites DataFrame\n",
    "        \"transcript_id\": result[\"transcript_id\"],\n",
    "        \"gene_id\": result[\"gene_id\"],\n",
    "        \"transcript_position\": result[\"transcript_position\"],\n",
    "        \"n_reads\": result[\"n_reads\"],\n",
    "        \"probability_modified\": result[\"probability_modified\"],\n",
    "        \"kmer\": result[\"kmer\"],\n",
    "        \"mod_ratio\": result[\"mod_ratio\"],\n",
    "        \"sample_id\": result[\"sample_id\"],\n",
    "        \"group_id\": result[\"group_id\"],\n",
    "        # Fields calculated by convert_to_genome_coordinates\n",
    "        \"genome_pos\": result[\"genome_pos\"],\n",
    "        \"chromosome\": result[\"chromosome\"],\n",
    "        \"dist_up_exon_junc\": result[\"dist_up_exon_junc\"],\n",
    "        \"dist_down_exon_junc\": result[\"dist_down_exon_junc\"],\n",
    "        \"region\": result[\"region\"]\n",
    "    }\n",
    "\n",
    "    # Append the reordered result to the results list\n",
    "    results.append(reordered_result)\n",
    "\n",
    "# Convert the final results list to a DataFrame\n",
    "annotated_modification_sites = pd.DataFrame(results)\n",
    "annotated_modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "48689f50-c3c2-49f9-b0d2-23a6f8afc14d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>n_reads</th>\n",
       "      <th>probability_modified</th>\n",
       "      <th>kmer</th>\n",
       "      <th>mod_ratio</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>group_id</th>\n",
       "      <th>genome_pos</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dist_up_exon_junc</th>\n",
       "      <th>dist_down_exon_junc</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1466</td>\n",
       "      <td>42</td>\n",
       "      <td>0.928761</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3222</td>\n",
       "      <td>78</td>\n",
       "      <td>0.955948</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1300627</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1320</td>\n",
       "      <td>157</td>\n",
       "      <td>0.936434</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.687898</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1223879</td>\n",
       "      <td>chr1</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360001</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1753</td>\n",
       "      <td>194</td>\n",
       "      <td>0.938970</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228561</td>\n",
       "      <td>chr1</td>\n",
       "      <td>386</td>\n",
       "      <td>92</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2612</td>\n",
       "      <td>31</td>\n",
       "      <td>0.952746</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.677419</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234848</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2612</td>\n",
       "      <td>192</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2579</td>\n",
       "      <td>27</td>\n",
       "      <td>0.995830</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234815</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2579</td>\n",
       "      <td>225</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000379198</td>\n",
       "      <td>ENSG00000176022</td>\n",
       "      <td>2432</td>\n",
       "      <td>22</td>\n",
       "      <td>0.987061</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1234668</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2432</td>\n",
       "      <td>372</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416718</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>341</td>\n",
       "      <td>801</td>\n",
       "      <td>0.927230</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>634716</td>\n",
       "      <td>chr1</td>\n",
       "      <td>341</td>\n",
       "      <td>205</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000465727</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1482</td>\n",
       "      <td>42</td>\n",
       "      <td>0.949604</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228941</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6</td>\n",
       "      <td>472</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000467278</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>2633</td>\n",
       "      <td>76</td>\n",
       "      <td>0.966780</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.868421</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1298659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>297</td>\n",
       "      <td>91</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1981</td>\n",
       "      <td>68</td>\n",
       "      <td>0.936208</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.779412</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219068</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1226</td>\n",
       "      <td>609</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000478938</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2414</td>\n",
       "      <td>97</td>\n",
       "      <td>0.911677</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.556701</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1659</td>\n",
       "      <td>176</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000492936</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>4747</td>\n",
       "      <td>61</td>\n",
       "      <td>0.956273</td>\n",
       "      <td>AGACA</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1304658</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1272</td>\n",
       "      <td>571</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2904</td>\n",
       "      <td>81</td>\n",
       "      <td>0.941420</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.765432</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1219381</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1971</td>\n",
       "      <td>611</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000494748</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>3337</td>\n",
       "      <td>111</td>\n",
       "      <td>0.903208</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>0.558559</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1218948</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2404</td>\n",
       "      <td>178</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6171</td>\n",
       "      <td>23</td>\n",
       "      <td>0.965541</td>\n",
       "      <td>AGACT</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>778545</td>\n",
       "      <td>chr1</td>\n",
       "      <td>82</td>\n",
       "      <td>260</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000506640</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>6083</td>\n",
       "      <td>29</td>\n",
       "      <td>0.980850</td>\n",
       "      <td>GAACT</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>s1</td>\n",
       "      <td>caud</td>\n",
       "      <td>774177</td>\n",
       "      <td>chr1</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000660930</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>1502</td>\n",
       "      <td>45</td>\n",
       "      <td>0.969710</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>s2</td>\n",
       "      <td>caud</td>\n",
       "      <td>1228927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>458</td>\n",
       "      <td>UTR3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id          gene_id  transcript_position  n_reads  \\\n",
       "7   ENST00000263741  ENSG00000078808                 1466       42   \n",
       "16  ENST00000354700  ENSG00000131584                 3222       78   \n",
       "3   ENST00000360001  ENSG00000078808                 1320      157   \n",
       "4   ENST00000360001  ENSG00000078808                 1753      194   \n",
       "14  ENST00000379198  ENSG00000176022                 2612       31   \n",
       "13  ENST00000379198  ENSG00000176022                 2579       27   \n",
       "12  ENST00000379198  ENSG00000176022                 2432       22   \n",
       "2   ENST00000416718  ENSG00000198744                  341      801   \n",
       "6   ENST00000465727  ENSG00000078808                 1482       42   \n",
       "17  ENST00000467278  ENSG00000131584                 2633       76   \n",
       "10  ENST00000478938  ENSG00000078808                 1981       68   \n",
       "11  ENST00000478938  ENSG00000078808                 2414       97   \n",
       "15  ENST00000492936  ENSG00000131584                 4747       61   \n",
       "8   ENST00000494748  ENSG00000078808                 2904       81   \n",
       "9   ENST00000494748  ENSG00000078808                 3337      111   \n",
       "1   ENST00000506640  ENSG00000228327                 6171       23   \n",
       "0   ENST00000506640  ENSG00000228327                 6083       29   \n",
       "5   ENST00000660930  ENSG00000078808                 1502       45   \n",
       "\n",
       "    probability_modified   kmer  mod_ratio sample_id group_id  genome_pos  \\\n",
       "7               0.928761  GGACT   0.761905        s2     caud     1228941   \n",
       "16              0.955948  AGACA   0.794872        s2     caud     1300627   \n",
       "3               0.936434  GGACT   0.687898        s1     caud     1223879   \n",
       "4               0.938970  AAACT   0.567010        s1     caud     1228561   \n",
       "14              0.952746  GGACT   0.677419        s2     caud     1234848   \n",
       "13              0.995830  GGACA   0.962963        s2     caud     1234815   \n",
       "12              0.987061  GGACT   0.909091        s2     caud     1234668   \n",
       "2               0.927230  GGACA   0.666667        s1     caud      634716   \n",
       "6               0.949604  GGACT   0.761905        s2     caud     1228941   \n",
       "17              0.966780  AGACA   0.868421        s2     caud     1298659   \n",
       "10              0.936208  GGACT   0.779412        s2     caud     1219068   \n",
       "11              0.911677  AAACT   0.556701        s2     caud     1218635   \n",
       "15              0.956273  AGACA   0.770492        s2     caud     1304658   \n",
       "8               0.941420  GGACT   0.765432        s2     caud     1219381   \n",
       "9               0.903208  AAACT   0.558559        s2     caud     1218948   \n",
       "1               0.965541  AGACT   0.913043        s1     caud      778545   \n",
       "0               0.980850  GAACT   0.655172        s1     caud      774177   \n",
       "5               0.969710  GGACT   0.777778        s2     caud     1228927   \n",
       "\n",
       "   chromosome  dist_up_exon_junc  dist_down_exon_junc region  \n",
       "7        chr1                  6                  472   UTR3  \n",
       "16       chr1                 66                  117   UTR3  \n",
       "3        chr1                 90                   46    CDS  \n",
       "4        chr1                386                   92   UTR3  \n",
       "14       chr1               2612                  192   UTR3  \n",
       "13       chr1               2579                  225   UTR3  \n",
       "12       chr1               2432                  372   UTR3  \n",
       "2        chr1                341                  205  ncRNA  \n",
       "6        chr1                  6                  472   UTR3  \n",
       "17       chr1                297                   91  ncRNA  \n",
       "10       chr1               1226                  609  ncRNA  \n",
       "11       chr1               1659                  176  ncRNA  \n",
       "15       chr1               1272                  571  ncRNA  \n",
       "8        chr1               1971                  611  ncRNA  \n",
       "9        chr1               2404                  178  ncRNA  \n",
       "1        chr1                 82                  260  ncRNA  \n",
       "0        chr1                104                    5  ncRNA  \n",
       "5        chr1                 20                  458   UTR3  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_modification_sites = annotated_modification_sites.sort_values(by='transcript_id')\n",
    "annotated_modification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0795c2af-6982-42e1-a46c-b595f7618ef9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotated modification sites saved to 3_test_output_test_data_m6anet_mod_sites.csv\n"
     ]
    }
   ],
   "source": [
    "# Save output\n",
    "output_prefix = '3_test_output'\n",
    "output_file = f\"{output_prefix}_{f}\"\n",
    "annotated_modification_sites.to_csv(output_file, index=False)\n",
    "\n",
    "print(\"Annotated modification sites saved to\", output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11c1f360-f74d-4f9f-bf41-ed652dfe65fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
